### Handling Imbalnced Dataset

In [2]:
# import the dataset

import pandas as pd
df = pd.read_csv('D:\Feature Engineering\Datasets\CreditCard Datasets\creditcard.csv')

In [3]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
df.isnull().sum() # There is no null values is present

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [5]:
df.shape

(284807, 31)

In [6]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

The Class feature have binary classification.Then see the data is imbalanced.

In [7]:
# Split the dataset into independent and dependent
X = df.drop('Class',axis = 1)
y = df.Class

In [8]:
X.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V20       V21       V22       V23       V24  \
0  0.098698  0.363787  ...  0.251412 -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425  ... -0.069083 -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  ...  0.524980  0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024  ... -0.208038 -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  ...  0.408542 -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  
0  0.128539 -0.189115  0.133558 -0.021053  149.62  
1  0.167170  0.125895 -0.008983  0.014724    2.69  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66  
3  0.647376 -0.221929  0.062723  0.061458  123.50  
4 -0.206010  0.502292  0.219422  0.215153   69.99  

[5 rows x 30 columns]

In [9]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Class, dtype: int64

In [10]:
# Perform train_test_split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,train_size=0.7)

In [11]:
print("X_train shape:", X_train.shape) # Ensure the both sample size are same
print("y_train shape:", y_train.shape)

X_train shape: (199364, 30)
y_train shape: (199364,)


In [12]:
# train the algorithm with imbalanced dataset and check accuracy
# taking logistic regression

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV
import numpy as np

In [13]:
10.0 ** np.arange(-2,3)

array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02])

In [14]:
logistic_reg = LogisticRegression()
# Perform Hyperparameter tunning
grid_param = {
    'C':10.0 ** np.arange(-2,3),
    'penalty':['l1','l2'] # Regularization Strength
}
cv = KFold(n_splits=5,random_state=None,shuffle=False)

In [15]:
gridcv = GridSearchCV(logistic_reg,grid_param,cv=cv,n_jobs=-1,scoring='f1_macro')
gridcv.fit(X_train,y_train)

D:\Anaconda\Anaconda\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
25 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "D:\Anaconda\Anaconda\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "D:\Anaconda\Anaconda\lib\site-packages\sklearn\linear_model\_logistic.py", line 1162, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "D:\Anaconda\Anaconda\lib\site-packages\sklearn\linear_model\_logistic.py", line 54, in _check_solver
    raise ValueError(
ValueError: Solver lb

GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]),
                         'penalty': ['l1', 'l2']},
             scoring='f1_macro')

In [16]:
gridcv.best_params_

{'C': 1.0, 'penalty': 'l2'}

In [17]:
y_pred = gridcv.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85252    45]
 [   43   103]]
0.9989700736163291
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85297
           1       0.70      0.71      0.70       146

    accuracy                           1.00     85443
   macro avg       0.85      0.85      0.85     85443
weighted avg       1.00      1.00      1.00     85443



The gridsearchcv is performed without error but giving warning messages.Because of not sufficient iterations

    We added the "solver" and "max_iter" hyperparameters to the grid search, allowing you to specify "liblinear" for the 
    "l1" penalty and increase the number of iterations.

    We also included "lbfgs" as a solver option for the "l2" penalty, as it is compatible with "l2."

By making these changes, you should be able to perform hyperparameter tuning without encountering the solver and convergence issues.

#### Use Some parameter and perform hyperparameter tunning for better results

The solver parameter specifies the algorithm to be used for optimization when fitting the logistic regression model. Different solvers use different optimization techniques to find the best set of coefficients (weights) for the logistic regression equation.

Here's a brief explanation of the two solver options you mentioned:

- liblinear Solver: 
    
    This is a good choice for small to medium-sized datasets. It is a solver that works well for both 
    'l1' and 'l2' penalties. It uses a coordinate descent algorithm for 'l1' penalty and a limited-memory Broyden-Fletcher-Goldfarb-Shanno (L-BFGS) optimization algorithm for 'l2' penalty. It's efficient and generally recommended for binary classification tasks.
    <br><br>

- lbfgs Solver: 

    This is another solver option that is suitable for 'l2' penalty (ridge regularization). It uses the Limited-memory Broyden-Fletcher-Goldfarb-Shanno (L-BFGS) optimization algorithm. 'lbfgs' tends to work well when you have a larger dataset compared to 'liblinear,' but it may require more iterations to converge.

The choice of solver can depend on various factors, including the size of your dataset, the type of regularization (penalty) you want to apply ('l1' or 'l2'), and the specific characteristics of your problem. It's a good practice to try multiple solvers and choose the one that performs best on your data. In GridSearchCV, you can include both 'liblinear' and 'lbfgs' in the solver options to find the best solver through hyperparameter tuning.

#### F1 Score

The F1 score is a commonly used metric for evaluating the performance of classification models, and the "macro" version of the F1 score, often referred to as "F1_macro," is one of its variants. It's used in situations where you want to assess the model's performance across multiple classes, and it provides a balanced measure of precision and recall that takes into account class imbalances.

In [18]:
# Perform again 
logistic_reg = LogisticRegression()
# Perform Hyperparameter tunning
grid_param = {
    'C':10.0 ** np.arange(-2,3),
    'penalty':['l1','l2'], # Regularization Strength
    'solver': ['liblinear', 'lbfgs'],  # Use 'liblinear' for 'l1' penalty
    'max_iter': [200,400,600,800,1000]  # Increase the number of iterations
}
cv = KFold(n_splits=5,random_state=None,shuffle=False)

In [19]:
gridcv = GridSearchCV(logistic_reg,grid_param,cv=cv,n_jobs=-1,scoring='f1_macro')
gridcv.fit(X_train,y_train)

D:\Anaconda\Anaconda\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
125 fits failed out of a total of 500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
125 fits failed with the following error:
Traceback (most recent call last):
  File "D:\Anaconda\Anaconda\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "D:\Anaconda\Anaconda\lib\site-packages\sklearn\linear_model\_logistic.py", line 1162, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "D:\Anaconda\Anaconda\lib\site-packages\sklearn\linear_model\_logistic.py", line 54, in _check_solver
    raise ValueError(
ValueError: Solver

GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]),
                         'max_iter': [200, 400, 600, 800, 1000],
                         'penalty': ['l1', 'l2'],
                         'solver': ['liblinear', 'lbfgs']},
             scoring='f1_macro')

In [20]:
gridcv.best_params_

{'C': 1.0, 'max_iter': 600, 'penalty': 'l1', 'solver': 'liblinear'}

In [21]:
y_pred = gridcv.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85278    19]
 [   51    95]]
0.9991807403766253
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85297
           1       0.83      0.65      0.73       146

    accuracy                           1.00     85443
   macro avg       0.92      0.83      0.87     85443
weighted avg       1.00      1.00      1.00     85443



using the both gridsearchcv the accuracy comes 99% but it is not correct accuracy for imbalanced dataset because FN and FP are the high amount.so we need to reduce them using other technique.

#### Just use random forest to check accuracy

In [22]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(X_train,y_train)

RandomForestClassifier()

In [23]:
y_pred = classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85291     6]
 [   27   119]]
0.9996137776061234
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85297
           1       0.95      0.82      0.88       146

    accuracy                           1.00     85443
   macro avg       0.98      0.91      0.94     85443
weighted avg       1.00      1.00      1.00     85443



Its amazing and nice use some hyperparameter tunning in random forest definitely the recall should increase.

class_weight = dict({0:1,1:100}) # one of the parameter of random forest used to increase the class count

example,

0    284315  <br>
1       492

0:1  284315 x 1 = 284315  <br>
1:100  492 x 100 = 49200
    
   Applying this in random forest and fit,finally gives the result.
   
   ##### NOTE : class weights also perform same as normal sometimes it performs better.It is based on the dataset

## Under Sampling

Under Sampling is a technique to reduce the points of the maximum labels
example,

10000 records to 100 like that

Under Sampling losses so much of data because of the count decrease

In [27]:
y_train.value_counts()  # to check value counts

0    199018
1       346
Name: Class, dtype: int64

In [36]:
# To perform under sampling

from collections import Counter
from imblearn.under_sampling import NearMiss
ns = NearMiss() # sampling_strategy is the percentage of samples taking
X_train_ns,y_train_ns = ns.fit_resample(X_train,y_train)

print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes before fit {}'.format(Counter(y_train_ns)))

The number of classes before fit Counter({0: 199018, 1: 346})
The number of classes before fit Counter({0: 346, 1: 346})


In [43]:
# Lets build the model using this under sampling

from sklearn.ensemble import RandomForestClassifier
classifiers = RandomForestClassifier()
classifiers.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [44]:
# Check for accuracy after predict
y_pred = classifiers.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[68377 16920]
 [   10   136]]
0.8018562082323888
              precision    recall  f1-score   support

           0       1.00      0.80      0.89     85297
           1       0.01      0.93      0.02       146

    accuracy                           0.80     85443
   macro avg       0.50      0.87      0.45     85443
weighted avg       1.00      0.80      0.89     85443



UnderSampling gives bad accuracy,recall and so on.It works best only for small datasets

## Over Sampling
Over Sampling is a technique to reduce the points of the minimum labels
example,

100 records to 10000 (Impute and increase the values)

In [50]:
# To perform Over sampling

from collections import Counter
from imblearn.over_sampling import RandomOverSampler
os = RandomOverSampler(sampling_strategy=0.75)   # sampling_strategy is the percentage of samples taking
X_train_os,y_train_os = os.fit_resample(X_train,y_train)

print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes before fit {}'.format(Counter(y_train_os)))

The number of classes before fit Counter({0: 199018, 1: 346})
The number of classes before fit Counter({0: 199018, 1: 149263})


In [51]:
# Lets build the model using this over sampling

from sklearn.ensemble import RandomForestClassifier
classifierss = RandomForestClassifier()
classifierss.fit(X_train_os,y_train_os)

RandomForestClassifier()

In [53]:
# Check for accuracy after predict
y_pred = classifierss.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85291     6]
 [   30   116]]
0.9995786664794073
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85297
           1       0.95      0.79      0.87       146

    accuracy                           1.00     85443
   macro avg       0.98      0.90      0.93     85443
weighted avg       1.00      1.00      1.00     85443



- Oversampling also reduce some errors.
- Overall randomforest works well,Include some hyperparameter tunning it works greater.
- Finally use one more technique to check it.

### SMOTE Tomek

SMOTE Tomek (SMOTE-Tomek) is a combination of two resampling techniques: Synthetic Minority Over-sampling Technique (SMOTE) and Tomek Links. It is used to address class imbalance in a dataset by oversampling the minority class and undersampling the majority class simultaneously.

In [55]:
# To perform SMOTETomek

from collections import Counter
from imblearn.combine import SMOTETomek
st = SMOTETomek(sampling_strategy=0.75) # sampling_strategy is the percentage of samples taking
X_train_st,y_train_st = st.fit_resample(X_train,y_train)

print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes before fit {}'.format(Counter(y_train_st)))

The number of classes before fit Counter({0: 199018, 1: 346})
The number of classes before fit Counter({0: 198303, 1: 148548})


In [57]:
# Lets build the model using this SMOTETomek

from sklearn.ensemble import RandomForestClassifier
classifierss = RandomForestClassifier()
classifierss.fit(X_train_st,y_train_st)

RandomForestClassifier()

In [58]:
# Check for accuracy after predict
y_pred = classifierss.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85279    18]
 [   25   121]]
0.9994967405170698
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85297
           1       0.87      0.83      0.85       146

    accuracy                           1.00     85443
   macro avg       0.94      0.91      0.92     85443
weighted avg       1.00      1.00      1.00     85443



the performance is good need more performance using hyperparameter tunning on random forest definetly gives more accuracy

### EasyEnsemble

EasyEnsemble is an ensemble learning technique designed to address class imbalance problems in classification tasks. It works by repeatedly undersampling the majority class and training multiple weak classifiers (typically decision trees) on these undersampled subsets. The final prediction is made by combining the predictions of all the weak classifiers.

In [64]:
from collections import Counter
from imblearn.ensemble import EasyEnsembleClassifier

# Create an EasyEnsemble classifier
easy_ensemble = EasyEnsembleClassifier(n_estimators=10, random_state=42)

# Fit the classifier to the training data
easy_ensemble.fit(X_train, y_train)

# Make predictions on the test data
y_pred = easy_ensemble.predict(X_test)

print('The number of classes before fit: {}'.format(Counter(y_train)))
print('The number of classes after fit: {}'.format(Counter(y_pred)))


The number of classes before fit: Counter({0: 199018, 1: 346})
The number of classes after fit: Counter({0: 81886, 1: 3557})


In [65]:
# Check for accuracy after predict
y_pred = easy_ensemble.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[81875  3422]
 [   11   135]]
0.9598211673279262
              precision    recall  f1-score   support

           0       1.00      0.96      0.98     85297
           1       0.04      0.92      0.07       146

    accuracy                           0.96     85443
   macro avg       0.52      0.94      0.53     85443
weighted avg       1.00      0.96      0.98     85443



It is bad performance because of ensemble techniques uses repeatedly undersampling the majority.

## CONCLUSION

- The main thing is that using each and every technique to handle imbalanced dataset
- Select best technique with selecting along the best classificaton algorithms like logistic,random forest etc,.
- Remember use hyperparameter tunning for algorithm to give better accuracy.(It is important)
- After analysis, apply algorithm and predict.
- Check the accuracy and show to stakeholders or get into next process.
